src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Tokens/Saml2SecurityTokenHandler.cs

```diff
@@ -0,0 +1,200 @@
+        /// <exception cref="ArgumentException">The token is not assignable from <see cref="Saml2SecurityToken"/>.</exception>
+        /// <exception cref="InvalidOperationException">Configuration <see cref="SecurityTokenHandlerConfiguration"/>is null.</exception>
+        /// <exception cref="ArgumentException">Saml2SecurityToken.Assertion is null.</exception>
+        /// <exception cref="SecurityTokenValidationException">Thrown if Saml2SecurityToken.Assertion.SigningToken is null.</exception>
+        /// <exception cref="SecurityTokenValidationException">Thrown if the certificate associated with the token issuer does not pass validation.</exception>
+        public override ReadOnlyCollection<ClaimsIdentity> ValidateToken(SecurityToken token)
+        {
+            Saml2SecurityToken samlToken = (Saml2SecurityToken)token;
+            string assertionXML = samlToken.AssertionXML;
+            SamlTokenValidationParameters tokenValidation = new SamlTokenValidationParameters();
+            ClaimsPrincipal claim = _internalSaml2SecurityTokenHandler.ValidateToken(assertionXML,
+                tokenValidation.ConvertToTokenValidationParameters(Configuration, samlToken), out MSIdentityTokens.SecurityToken msIdentityToken);
+            return claim.Identities?.ToList().AsReadOnly();
+        }
+        #endregion
+
+        #region TokenSerialization
+
+        /// <summary>
+        /// Indicates whether the current XML element can be read as a token 
+        /// of the type handled by this instance.
+        /// </summary>
+        /// <param name="reader">An XML reader positioned at a start 
+        /// element. The reader should not be advanced.</param>
+        /// <returns>'True' if the ReadToken method can the element.</returns>
+        public override bool CanReadToken(XmlReader reader)
+        {
+            return _internalSaml2SecurityTokenHandler.CanReadToken(reader);
+        }
+
+        /// <summary>
+        /// Deserializes from XML a token of the type handled by this instance.
+        /// </summary>
+        /// <param name="reader">An XML reader positioned at the token's start 
+        /// element.</param>
+        /// <returns>An instance of <see cref="Saml2SecurityToken"/>.</returns>
+        /// <exception cref="InvalidOperationException">Is thrown if 'Configuration' or 'Configruation.IssuerTokenResolver' is null.</exception>
+        public override SecurityToken ReadToken(XmlReader reader)
+        {
+            XmlDocument doc = new XmlDocument();
+            XmlElement rstXml = (doc.ReadNode(reader) as XmlElement);
+            MSIdentityTokens.Saml2.Saml2SecurityToken internalSecurityToken = (MSIdentityTokens.Saml2.Saml2SecurityToken)_internalSaml2SecurityTokenHandler.ReadToken(rstXml.OuterXml);
+            TryResolveIssuerToken(internalSecurityToken.Assertion, Configuration.IssuerTokenResolver, out SecurityToken token);
```

Comment:
Use Saml2SecurityTokenHandler.ReadSaml2Token to avoid the cast. Method [here](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/dev/src/Microsoft.IdentityModel.Tokens.Saml/Saml2/Saml2SecurityTokenHandler.cs#L553)